File: ios/Hairdressing/Hairdressing/Services/PriceService.swift
//
//  PriceService.swift
//  Hairdressing
//
//  Created by Laszlo Losonczi on 2025. 11. 28..
//

import Foundation
import Combine

final class PriceService: ObservableObject {
  private let httpClient: HttpClient

  init (httpClient: HttpClient) {
    self.httpClient = httpClient
  }

  @Published var prices: [Price.Response] = []

  func listPrices() async throws -> Void {
    let result = try await httpClient.get("/price/v1/prices", responseType: [Price.Response].self)
    prices = result
  }

  func createPrice(body: Price.Request) async throws -> Void {
    let result = try await httpClient.post("/price/v1/prices", body: body, responseType: Price.PriceId.self);

    prices.append(
        Price
          .Response(
            priceId: result.priceId,
            name: body.name,
            amount: body.amount,
            unitOfMeasurement: body.unitOfMeasurement
          )
    )
    prices.sort(using: KeyPathComparator(\.name))
  }

  func updatePrice(priceId: Price.Id, body: Price.Request) async throws -> Void {
    let result = try await httpClient.put("/price/v1/prices/\(priceId)", body: body, responseType: Price.PriceId.self);

    if let index =  prices.firstIndex(where: {price in
      price.priceId == result.priceId
    }) {
      prices[index].name = body.name
      prices[index].amount = body.amount
      prices[index].unitOfMeasurement = body.unitOfMeasurement
    }

    prices.sort(using: KeyPathComparator(\.name))
  }

  func deletePrice(priceId: Price.Id) async throws {
    try await httpClient.delete("/price/v1/prices/\(priceId)");

    prices.removeAll{
      $0.priceId == priceId
    }
  }
}

File: web/projects/hairdressing/src/services/price-service.ts
import { HttpClient } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { API_URL } from '@household/shared-ui';
import { Api } from '@household/shared/types/api';
import { Requests } from '@household/shared/types/requests';
import { Responses } from '@household/shared/types/responses';

@Injectable({
  providedIn: 'root',
})
export class PriceService {
  private httpClient = inject(HttpClient);
  private apiUrl = inject(API_URL);

  listPrices() {
    return this.httpClient.get<Responses.Price[]>(`${this.apiUrl}/price/v1/prices`);
  }
  
  createPrice(body: Requests.Price) {
    return this.httpClient.post<Api.Price.PriceId>(`${this.apiUrl}/price/v1/prices`, body);
  }
  
  updatePrice(priceId: Api.Price.Id, body: Requests.Price) {
    return this.httpClient.put<Api.Price.PriceId>(`${this.apiUrl}/price/v1/prices/${priceId}`, body);
  }
  
  deletePrice(priceId: Api.Price.Id) {
    return this.httpClient.delete(`${this.apiUrl}/price/v1/prices/${priceId}`);
  }
}
